src/pcp/ss/pcp-ss.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2021,2023 Red Hat.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PCP-SS 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pcp-ss\f1 \- report socket statistics
.SH SYNOPSIS
\f3pcp\f1 [\f2pcp\ options\f1] \f3ss\f1 [\f2ss\ options\f1]
.SH DESCRIPTION
.B pcp-ss
reports socket statistics collected by the
.BR pmdasockets (1)
PMDA agent.
The command is intended to be reasonably compatible with many of the
.BR ss (8)
command line options and reporting formats, but also offer
the advantages of local or remote monitoring (in live mode) and
also historical replay from a previously recorded PCP archive.
Note that since
.BR ss (1)
has many command line options, many of which are the same as standard PCP command line options as described in
.BR PCPIntro (1),
the
.B pcp-ss
tool should always be invoked by users
using the
.B pcp
front-end.
This allows standard PCP commandline options such as
.BR \-h ,
.BR \-a ,
.BR \-S ,
.BR \-T ,
.BR \-O ,
.BR \-z ,
etc to be passed without conflict with
.BR ss (1)
options.
See the
.B EXAMPLES
sections below for typical usage and command lines.
.PP
Live mode uses the
.B pcp
\fB-h\fP \fIhost\fP option and requires the
.BR pmdasockets (1)
PMDA to be installed and enabled on the target \fIhost\fP (local or remote), see
.BR pmdasockets (1)
for details on how to enable the \fBsockets\fP PMDA on a particular host.
The default source is live metrics collected on
.BR localhost ,
if neither of the
.B \-h
or
.B \-a
options are given.
.PP
Historical/archive replay uses the
.B pcp
\fB-a\fP \fIarchive\fP option, where \fIarchive\fP is the
basename of a previously recorded PCP archive.
The archive replay feature is particularly useful because
socket statistics can be reported for a designated time using the
.B pcp
.B \-\-origin
option (which defaults to the start time of the archive).
.SH EXAMPLES
.TP 5
\fBpcp ss\fP
Display default basic socket information for the local host.
This includes \fBNetid\fP (\fBtcp\fP, \fBudp\fP, etc), \fBState\fP (\fBESTAB\fP,
\fBTIME_WAIT\fP, etc), \fBRecv-Q\fP and \fBSend-Q\fP queue lengths
and the local and peer address and port for each socket.
.TP 5
\fBpcp \-h \fIsomehost\fP ss \-noemitauO\fP
Display the same basic socket information as above for the host \fIsomehost\fP,
which may be the default \fBlocalhost\fP.
The additional command line arguments (\fB\-noemitauO\fP) display
one line per socket (\fB\-O\fP), numeric (\fB\-n\fP) service names (default),
timer information (\fB\-o\fP), extended socket details (\fB\-e\fP),
socket memory usage (\fB\-m\fP), internal TCP information (\fB\-i\fP),
both udp (\fB\-u\fP) and tcp sockets (\fB\-t\fP) and both listening and
non-listening sockets (\fB\-a\fP).
.TP 5
\fBpcp \-a \fIsomearchive\fP \fB-S'@Wed 16 Jun 2021 12:57:21'\fP ss \-noemitauO\fP
Display the same information as the above example, but for the archive
\fIsomearchive\fP starting at the given time \fBWed 16 Jun 2021 12:57:21\fP.
Note the literal \fB@\fP prefix is required for an absolute time, see
.BR PCPIntro (1)
for details.
The archive must of course contain data for the requested time. You can use
\fBpmlogdump \-l\fP \fIsomearchive\fP to examine the time bounds of \fIsomearchive\fP.
.TP 5
\fBpcp \-a \fIsomearchive\fP \fB\-O\-0\fP ss \-noemitauO\fP
As above, but with an offset of zero seconds (\fB\-O\-0\fP) before the current end of
\fIsomearchive\fP, i.e. the most recently logged data. Note that \fIsomearchive\fP
may be currently growing (i.e. being logged with
.BR pmlogger (1)).
.SH OPTIONS
Due to the large number of options supported by
.BR pcp-ss ,
the
.BR pcp (1)
command should always be used to invoke
.B pcp-ss
in order to specify options such as the metrics source (host or archive)
and also (in archive mode), the requested start time or offset, and timezone
using the following options:
.TP 5
\fB\-h\fP, \fB\-\-host\fP
The remote hostname to connect to in live mode.
.TP 5
\fB\-a\fP, \fB\-\-archive\fP
The archive file to use for historical sampling
.TP 5
\fB\-O\fP, \fB\-\-origin\fP
The time offset to use within an archive (implies
.BR \-a )
.TP 5
\fB\-S\fP, \fB\-\-start\fP
The start time (e.g. in
.BR ctime (3)
format) to use when replaying an archive.
.TP 5
\fB\-Z\fP, \fB\-\-timezone\fP
Use a specific timezone.
Since
.B pcp-ss
doesn't report timestamps, this only affects the interpretation
of an absolute starting time (\fB\-S\fP) or offset (\fB\-O\fP).
.TP 5
\fB\-z\fP, \fB\-\-hostzone\fP
In archive mode, use the timezone of the archive rather than the
timezone on the local machine running
.BR pcp-ss .
The timezone, start and finish times of the archive may be examined using
.BR pmlogdump (1)
with the \fB\-L\fP option.
.PP
The above
.B pcp
options become indirectly available to the
.B pcp-ss
command via environment variables - refer to
.BR PCPIntro (1)
for a complete description of these options.
.PP
The additional command line options available for
.B pcp-ss
itself are:
.TP 5
\fB\-h\fP, \fB\-\-help\fP
show help message and exit
.TP 5
\fB\-V\fP, \fB\-\-version\fP
output version information
.TP 5
\fB\-n\fP, \fB\-\-numeric\fP
don't resolve service names (currently always set)
'\" .TP 5
'\" \fB\-r\fP, \fB\-\-resolve\fP
'\" resolve host names (not yet implemented)
.TP 5
\fB\-a\fP, \fB\-\-all\fP
display all sockets
.TP 5
\fB\-l\fP, \fB\-\-listening\fP
display listening sockets
.TP 5
\fB\-o\fP, \fB\-\-options\fP
show timer information
.TP 5
\fB\-e\fP, \fB\-\-extended\fP
show detailed socket information
.TP 5
\fB\-m\fP, \fB\-\-memory\fP
show socket memory usage
'\" .TP 5
'\" \fB\-p\fP, \fB\-\-processes\fP
'\" show process using socket
.TP 5
\fB\-i\fP, \fB\-\-info\fP
show internal TCP information
'\" .TP 5
'\" \fB\-s\fP, \fB\-\-summary\fP
'\" show socket usage summary
'\" .TP 5
'\" \fB\-b\fP, \fB\-\-bpf\fP
'\" show bpf filter socket information
'\" .TP 5
'\" \fB\-E\fP, \fB\-\-events\fP
'\" continually display sockets as they are destroyed
'\" .TP 5
'\" \fB\-Z\fP, \fB\-\-context\fP
'\" display process SELinux security contexts
'\" .TP 5
'\" \fB\-z\fP, \fB\-\-contexts\fP
'\" display process and socket SELinux security contexts
'\" .TP 5
'\" \fB\-N\fP, \fB\-\-net\fP
'\" switch to the specified network namespace name
.TP 5
\fB\-4\fP, \fB\-\-ipv4\fP
display only IP version 4 sockets
.TP 5
\fB\-6\fP, \fB\-\-ipv6\fP
display only IP version 6 sockets
'\" .TP 5
'\" \fB\-0\fP, \fB\-\-packet\fP
'\" display PACKET sockets
.TP 5
\fB\-t\fP, \fB\-\-tcp\fP
display only TCP sockets
'\" .TP 5
'\" \fB\-M\fP, \fB\-\-mptcp\fP
'\" display only MPTCP sockets
'\" .TP 5
'\" \fB\-S\fP, \fB\-\-sctp\fP
'\" display only SCTP sockets
.TP 5
\fB\-u\fP, \fB\-\-udp\fP
display only UDP sockets
'\" .TP 5
'\" \fB\-d\fP, \fB\-\-dccp\fP
'\" display only DCCP sockets
'\" .TP 5
'\" \fB\-w\fP, \fB\-\-raw\fP
'\" display only RAW sockets
'\" .TP 5
'\" \fB\-x\fP, \fB\-\-unix\fP
'\" display only Unix domain sockets
.TP 5
\fB\-H\fP, \fB\-\-noheader\fP
Suppress header line
.TP 5
\fB\-O\fP, \fB\-\-oneline\fP
socket's data printed on a single line
.SH REPORT
The columns in the
.B pcp-ss
report vary according to the command line options and have the
same interpretation as described in
.BR ss (8).
.PP
One difference with
.B pcp-ss
is that the first line in the report begins with '\fB# Timestamp\fP'
followed by the timestamp (in the requested timezone, see \fB\-z\fP and \fB\-Z\fP above)
of the sample data from the host or archive source.
Following the timestamp is the currently active filter string for the metrics source.
In archive mode, the active filter can be changed dynamically, even whilst the archive is
being recorded.
This is different to
.BR ss (8)
where the filter is optionally specified on the command line of the tool and is
always 'live', i.e.
.BR ss (8)
does not support retrospective replay.
With
.BR pcp-ss ,
the filter is stored in the back-end PMDA, see
.BR pmdasockets (1),
in the metric
.BR network.persocket.filter.
The default filter is \fBstate connected\fP, which can be changed by
storing a new string value in the
.B network.persocket.filter
metric using
.BR pmstore (1),
e.g.
\fBpmstore network.persocket.filter "state established"\fP.
This will override the persistent default filter, which is stored
in a PMDA configuration file and loaded each time the sockets PMDA is started.
See
.BR pmdasockets (1)
for further details and see
.BR ss (8)
for details of the filter syntax and examples.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.PP
For environment variables affecting PCP tools, see \fBpmGetOptions\fP(3).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pcp (1),
.BR pmdasockets (1),
.BR pmlogger (1),
.BR pcp.conf (5)
and
.BR ss (8).

.\" control lines for scripts/man-spell
.\" +ok+ somearchive commandline TIME_WAIT persocket
.\" +ok+ noemitauO {from command line option} somehost ESTAB Netid Recv Jun


CHANGELOG
pcp-6.3.6 (17 March 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.4 (13 March 2025)
    - libpcp: fix endian issue in big-endian __pmTimestamp writing
    - pcp-xsos: fine-tune error handling on bad command line options
    - pcp-xsos: single-line pmPrintValue/pminfo values, escape shell chars
    - pcp-atop: correct BPF per-process network metric names
    - pmcd, pmdaproc: allow non-local username connection attributes
    - pmlogger, libpcp: fix handling of dynamic instance domain labels
    - pmrep: use BPF instead of BCC for proc networking metrics
    - pmlogconf: use libbpf over BCC for networking metrics
    - pmlogextract: fix logic around instances labels, allowing set updates
    - pmdalinux: mem.numa.hugepages metrics (per node, per hugepage size)
    - pmdabpftrace: fix for some python3 versions async IO create_task
    - pmdanvidia: if present use dso from the libnvidia-ml runtime package
    - pmdadm, pmieconf: default to enabling the dmthin pmie rules
    - configure.ac: fix autoreconf, add AC_DEFINE_UNQUOTED descriptions
    - rpm builds: enable bpf, bpftrace, bcc on riscv64 builds
    - rpm builds: drop user creation scriptlets for rpm builds
    - deb build: force debian/rules to be recreated each time
    - containers: updates to the archive-analysis container
    - selinux: several policy updates for new AVC denials
    - docs: improve SSL/TLS diagnostic in pcp-verify script

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out

src/pcp/ss/pcp-ss.py
#!/usr/bin/env pmpython
#
# Copyright (C) 2021 Red Hat.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# pylint: disable=line-too-long,bad-continuation,broad-except,bare-except
# pylint: disable=missing-docstring,multiple-imports,unused-import
# pylint: disable=too-many-boolean-expressions,consider-using-dict-items

""" Display socket statistics """

import os, sys, argparse, errno

# PCP Python PMAPI
from pcp import pmapi
from cpmapi import PM_CONTEXT_ARCHIVE, PM_CONTEXT_HOST
from cpmapi import PM_ERR_EOL, PM_MODE_INTERP

# pmns prefix for pmdasockets(1) metrics
pmns = "network.persocket"

def remove_prefix(text, prefix):
    """ remove prefix from text (for python < 3.9) """
    return text[text.startswith(prefix) and len(prefix):]

class SS(object):
    """ PCP client implementation of the ss(1) tool, using pmdasockets(1) """

    def __init__(self):
        """ Construct object - prepare for command line handling """
        self.args = self.options()
        self.source = "local:"
        self.speclocal = None
        self.metrics = []
        self.instD = {} # { instid: instname }
        self.metricsD = {} # { name: fetchgroup }
        self.descsD = {} # { name: pmDesc }
        self.valuesD = {} # { name: {instid: value} }
        self.context = None
        self.pmfg = None

    def options(self):
        """ define command line arguments """
        p = argparse.ArgumentParser()
        p.add_argument('-V', '--version', action='store_true', help='output version information')
        p.add_argument('-n', '--numeric', action='store_true', help='don\'t resolve service names or port names (currently always set)')
        #p.add_argument('-r', '--resolve', action='store_true', help='resolve host names (currently never set)')
        p.add_argument('-a', '--all', action='store_true', help='display both listening and non-listening states')
        p.add_argument('-l', '--listening', action='store_true', default=False, help='display only listening sockets')
        p.add_argument('-o', '--options', action='store_true', help='show timer information')
        p.add_argument('-e', '--extended', action='store_true', help='show detailed socket information')
        p.add_argument('-m', '--memory', action='store_true', help='show socket memory usage')
        #p.add_argument('-p', '--processes', action='store_true', help='show process using socket (not yet implemented)')
        p.add_argument('-i', '--info', action='store_true', help='show internal TCP information')
        #p.add_argument('-s', '--summary', action='store_true', help='show socket usage summary (not yet implemented)')
        #p.add_argument('-b', '--bpf', action='store_true', help='show bpf filter socket information (not yet implemented)')
        #p.add_argument('-E', '--events', action='store_true', help='continually display sockets as they are destroyed (not implemented)')
        #p.add_argument('-Z', '--context', action='store_true', help='display process SELinux security contexts (not implemented)')
        #p.add_argument('-z', '--contexts', action='store_true', help='display process and socket SELinux security contexts (not implemented)')
        #p.add_argument('-N', '--net', action='store_true', help='switch to the specified network namespace name (not implemented)')
        p.add_argument('-4', '--ipv4', action='store_true', help='display only IP version 4 sockets')
        p.add_argument('-6', '--ipv6', action='store_true', help='display only IP version 6 sockets')
        #p.add_argument('-0', '--packet', action='store_true', help='display PACKET sockets (not implemented)')
        p.add_argument('-t', '--tcp', action='store_true', help='display only TCP sockets')
        #p.add_argument('-M', '--mptcp', action='store_true', help='display only MPTCP sockets (not implemented)')
        #p.add_argument('-S', '--sctp', action='store_true', help='display only SCTP sockets (not implemented)')
        p.add_argument('-u', '--udp', action='store_true', help='display only UDP sockets')
        #p.add_argument('-d', '--dccp', action='store_true', help='display only DCCP sockets (not implemented)')
        #p.add_argument('-w', '--raw', action='store_true', help='display only RAW sockets (not implemented)')
        #p.add_argument('-x', '--unix', action='store_true', help='display only Unix domain sockets (not implemented)')
        p.add_argument('-H', '--noheader', action='store_true', help='Suppress header line')
        p.add_argument('-O', '--oneline', action='store_true', help='print each socket\'s data on a single line')
        args = p.parse_args()

        # special cases
        if not (args.tcp or args.udp): # or args.mptcp or args.sctp or args.packet or args.dccp or args.raw or args.unix):
            args.tcp = args.udp = True # args.mptcp = args.sctp = args.packet = args.dccp = args.raw = args.unix = True

        if not (args.ipv4 or args.ipv6):
            # default to both ipv4 and ipv6, subject to the prevailing filter
            args.ipv4 = args.ipv6 = True

        return args

    def connect(self):
        """ Establish a fetchgroup PMAPI context to archive, host or local,
            via environment passed in by pcp(1). The ss(1) command has many
            clashes with standard PCP arguments so this is the only supported
            invocation. Debug options (if any) are set for us by pcp(1).
            Return True or False if we fail to connect.
        """
        # source
        pcp_host = os.getenv("PCP_HOST")
        pcp_archive = os.getenv("PCP_ARCHIVE")

        # time window - only via environment: too many clashes with ss args
        pcp_origin = os.getenv("PCP_ORIGIN_TIME")
        pcp_start_time = os.getenv("PCP_START_TIME")
        pcp_align_time = os.getenv("PCP_ALIGN_TIME")
        pcp_timezone = os.getenv("PCP_TIMEZONE")
        pcp_hostzone = os.getenv("PCP_HOSTZONE")
        pcp_debug = os.getenv("PCP_DEBUG")

        if pcp_archive is not None:
            self.context_type = PM_CONTEXT_ARCHIVE
            if pcp_origin is None and pcp_start_time is None:
                pcp_origin = "-0" # end of archive
            self.source = pcp_archive
        else:
            self.context_type = PM_CONTEXT_HOST
            if pcp_host is not None:
                self.source = pcp_host
            else:
                self.source = "localhost"

        try:
            self.pmfg = pmapi.fetchgroup(self.context_type, self.source)
            self.context = self.pmfg.get_context()
            if pcp_archive:
                options = pmapi.pmOptions("a:A:O:S:D:zZ:")
                optargv = ["pcp-ss", pcp_archive]
                if pcp_debug:
                    optargv.append("-D%s" % pcp_debug)
                if pcp_align_time:
                    optargv.append("-A%s" % pcp_align_time)
                if pcp_timezone:
                    optargv.append("-Z%s" % pcp_timezone)
                if pcp_hostzone:
                    optargv.append("-z")
                if pcp_origin:
                    optargv.append("-O%s" % pcp_origin)
                    pmapi.pmContext.fromOptions(options, optargv)
                    origin = options.pmGetOptionOrigin()
                    self.context.pmSetMode(PM_MODE_INTERP, origin, 0)
                elif pcp_start_time:
                    optargv.append("-S%s" % pcp_start_time)
                    pmapi.pmContext.fromOptions(options, optargv)
                    start = options.pmGetOptionStart()
                    self.context.pmSetMode(PM_MODE_INTERP, start, 0)

        except pmapi.pmErr as pmerr:
            sys.stderr.write("%s: %s '%s'\n" % (pmerr.progname(), pmerr.message(), self.source))
            return False

        # check network.persocket metrics are available
        try:
            self.context.pmLookupName((pmns + ".filter"))
        except Exception:
            if self.context_type == PM_CONTEXT_HOST:
                msg = "on host %s.\nIs the 'sockets' PMDA installed and enabled? See pmdasockets(1)." % self.source
            else:
                msg = "in archive %s" % self.source
            print("Error: metrics for '%s' not found %s" % (pmns, msg))
            return False

        return True

    def traverseCB(self, name):
        if not name.endswith(".filter"):
            self.metrics.append(name)

    def fetch(self):
        """ fetch metrics and report as per given options """

        # filter and timestamp
        self.filter = self.pmfg.extend_item(pmns + ".filter")
        self.timestamp = self.pmfg.extend_timestamp()

        self.context.pmTraversePMNS(pmns, self.traverseCB)
        self.pmids = self.context.pmLookupName(self.metrics)
        descs = self.context.pmLookupDescs(self.pmids)

        # Create metrics dict keyed by metric name (without pmns prefix).
        nmetrics = len(self.metrics)
        for i in range (0, nmetrics):
            pmnsname = self.metrics[i]
            name = remove_prefix(pmnsname, pmns + ".")
            try:
                # do not want rate conversion, so use "instant" scale and only one fetch
                self.descsD[name] = descs[i]
                self.metricsD[name] = self.pmfg.extend_indom(pmnsname,
                    descs[i].contents.type, scale="instant", maxnum=10000)
            except Exception as e:
                print("Warning: Failed to add %s to fetch group: %s" % (name, e))

        # fetch the lot
        try:
            self.pmfg.fetch()
        except Exception as e:
            print("Error: fetch failed: %s" % e)
            sys.exit(1)

        # extract instances and values
        for name in self.metricsD:
            try:
                instvalsD = {}
                # walk the instances in the fetch group for this metric
                for inst, iname, value in self.metricsD[name]():
                    self.instD[inst] = iname
                    try:
                        instvalsD[inst] = value()
                    except Exception as e:
                        print("Error: value() failed for metric %s, inst %d, iname %s: %s" % (name, inst, iname, e))
                self.valuesD[name] = instvalsD
            except Exception as e:
                pass # instance went away, socket probably closed

    def strfield(self, fmt, metric, inst, default=""):
        """ return formatted field, if metric and inst are available else default string """
        try:
            s = self.valuesD[metric][inst]
            if s is not None:
                return fmt % s
        except Exception:
            pass
        return fmt % default

    def intfield(self, fmt, metric, inst, default=0):
        """ return formatted field, if metric and inst are available else default """
        try:
            s = self.valuesD[metric][inst]
            if s is not None:
                return fmt % s
        except Exception:
            pass
        return fmt % default

    def boolfield(self, field, metric, inst, default=""):
        """ return field if metric and inst are available and non-zero """
        try:
            s = self.valuesD[metric][inst]
            if s is not None and s != 0:
                return field
        except Exception:
            pass
        return default

    def filter_netid(self, inst):
        """ filter on netid and -t, -u and -x cmdline options """
        ret = False
        netid = self.valuesD["netid"][inst]
        if self.args.tcp and netid == "tcp":
            ret = True
        elif self.args.udp and netid == "udp":
            state = self.valuesD["state"][inst]
            ret = bool(state != "UNCONN" or self.args.listening)
        #elif self.args.unix and netid == "unix":
        #    ret = True
        #elif self.args.raw and netid == "raw":
        #    ret = True
        return ret

    def filter_ipv46(self, inst):
        """ filter on ip v4 or v6 """
        ret = False
        if self.valuesD["src"][inst]:
            v6 = self.valuesD["src"][inst][0] == '['
        elif self.valuesD["dst"][inst]:
            v6 = self.valuesD["dst"][inst][0] == '['
        else:
            v6 = False
        if self.args.ipv6 and v6:
            ret = True
        if self.args.ipv4 and not v6:
            ret = True
        return ret

    def filter_listening(self, inst):
        """ filter on tcp socket listening state """
        if self.args.all:
            return True
        state = self.valuesD["state"][inst]
        if self.args.listening:
            netid = self.valuesD["netid"][inst]
            if state == "UNCONN" and netid == 'udp':
                return True
            if state != "LISTEN":
                return False
        else:
            if state == "LISTEN":
                return False
        return True

    def report(self):
        """ output report based on cmdline options """
        if not self.args.noheader: # -H flag
            print("# Time: %s Filter: %s" % (self.timestamp(), self.filter()))
            print("Netid  State  Recv-Q Send-Q %25s %-25s  Process" % ("Local Address:Port", "Peer Address:Port"))
        for inst in self.instD:
            if not self.filter_netid(inst) or not self.filter_listening(inst):
                continue
            if not self.filter_ipv46(inst):
                continue
            out = ""
            out += self.strfield("%-6s", "netid", inst)
            out += self.strfield("%-9s", "state", inst, "-")
            out += self.strfield("%6u", "recvq", inst, 0)
            out += self.strfield("%6u", "sendq", inst, 0)
            out += self.strfield(" %25s", "src", inst)
            out += self.strfield(" %-25s", "dst", inst)

            if self.args.options: # -o --options flag
                m = self.valuesD["timer.str"][inst]
                if m is not None and len(m) > 0:
                    out += " timer(%s)" % m

            if self.args.extended: # -e --extended flag
                out += self.strfield(" uid:%d", "uid", inst, 0)
                out += self.strfield(" inode:%lu", "inode", inst, 0)
                out += self.strfield(" sk:%x", "sk", inst, 0)
                out += self.strfield(" cgroup:%s", "cgroup", inst)
                if self.valuesD["v6only"][inst] != 0:
                    out += " v6only:%d" % self.valuesD["v6only"][inst]
                out += " <->"

            if not self.args.oneline and (self.args.memory or self.args.info):
                out += "\n"

            if self.args.memory: # -m --memory flag
                m = self.valuesD["skmem.str"][inst]
                if m is not None and len(m) > 0:
                    out += " skmem(%s)" % m

            if self.args.info: # -i --info flag
                out += self.boolfield(" ts", "ts", inst)
                out += self.boolfield(" sack", "sack", inst)
                # TODO ecn, ecnseen, fastopen
                out += self.boolfield(" cubic", "cubic", inst)
                out += self.strfield(" wscale:%s", "wscale.str", inst)
                # TODO rto, backoff
                out += self.strfield(" rtt:%s", "round_trip.str", inst)
                out += self.strfield(" ato:%.0lf", "ato", inst, 0.0)
                out += self.strfield(" mss:%d", "mss", inst, 0)
                out += self.strfield(" cwnd:%d", "cwnd", inst, 0)
                out += self.strfield(" pmtu:%d", "pmtu", inst, 0)
                out += self.strfield(" ssthresh:%d", "ssthresh", inst, 0)
                out += self.strfield(" bytes_sent:%lu", "bytes_sent", inst, 0)
                out += self.strfield(" bytes_acked:%lu", "bytes_acked", inst, 0)
                out += self.strfield(" bytes_received:%lu", "bytes_received", inst, 0)
                out += self.strfield(" segs_out:%lu", "segs_out", inst, 0)
                out += self.strfield(" segs_in:%lu", "segs_in", inst, 0)
                out += self.strfield(" send %.0lfbps", "send", inst, 0)
                out += self.strfield(" lastsnd:%.0lf", "lastsnd", inst, 0.0)
                out += self.strfield(" lastrcv:%.0lf", "lastrcv", inst, 0.0)
                out += self.strfield(" lastack:%.0lf", "lastack", inst, 0.0)
                out += self.strfield(" pacing_rate %.0lfbps", "pacing_rate", inst, 0.0)
                out += self.strfield(" delivery_rate %.0lfbps", "delivery_rate", inst, 0.0)
                # TODO max_pacing_rate
                out += self.strfield(" rcv_space:%lu", "rcv_space", inst, 0)
                # TODO tcp-ulp-mptcp
                # TODO token:<rem_token(rem_id)/loc_token(loc_id)>
                # TODO seq:<sn>
                # TODO sfseq:<ssn>
                # TODO ssnoff

            print(out)

if __name__ == '__main__':
    try:
        ss = SS()
        if not ss.connect():
            # failed to connect or metrics not found - error already reported.
            sys.exit(1)
        ss.fetch()
        ss.report()
    except pmapi.pmErr as error:
        if error.args[0] == PM_ERR_EOL:
            sys.exit(0)
        sys.stderr.write('%s: %s\n' % (error.progname(), error.message()))
        sys.exit(1)
    except pmapi.pmUsageErr as usage:
        usage.message()
        sys.exit(1)
    except IOError as error:
        if error.errno != errno.EPIPE:
            sys.stderr.write("Error: %s\n" % str(error))
            sys.exit(1)
    except KeyboardInterrupt:
        sys.stdout.write("\n")
